servers/zms/src/main/java/com/yahoo/athenz/zms/ZMSImpl.java

```diff
@@ -1543,24 +1555,60 @@ void deleteDomain(ResourceContext ctx, String auditRef, String domainName, Strin
             groupMemberConsistencyCheck(domainName, group.getName(), true, caller);
         }
 
-        // consistency checks are ok. Now make sure no service is dependent on the domain
+        // consistency checks are ok. Now check if there are provider services dependent on the domain
 
-        verifyNoServiceDependenciesOnDomain(domainName, caller);
+        ServiceIdentityList serviceIdentityList = dbService.listServiceDependencies(domainName);
+        if (serviceIdentityList.getNames() != null && !serviceIdentityList.getNames().isEmpty()) {
+            tryRemoveServiceDependenciesForDomain(ctx, auditRef, domainName, caller, serviceIdentityList);
+        }
 
         // no service is dependent on the domain, we can go ahead and delete the domain
 
         dbService.executeDeleteDomain(ctx, domainName, auditRef, caller);
     }
 
-    private void verifyNoServiceDependenciesOnDomain(String domainName, String caller) {
-        ServiceIdentityList serviceIdentityList = dbService.listServiceDependencies(domainName);
-        if (serviceIdentityList.getNames() != null && !serviceIdentityList.getNames().isEmpty()) {
-            StringBuilder msgBuilder = new StringBuilder("Remove domain '");
-            msgBuilder.append(domainName);
-            msgBuilder.append("' dependency from the following service(s):");
-            msgBuilder.append(String.join(", ", serviceIdentityList.getNames()));
-            throw ZMSUtils.forbiddenError(msgBuilder.toString(), caller);
+    private void tryRemoveServiceDependenciesForDomain(ResourceContext ctx, String auditRef, String domainName, String caller, ServiceIdentityList serviceIdentityList) {
```

Comment:
I'm not sure if we're following the design here. There are two cases that need to be addressed here:
a) services that are using our dependency / tenancy api - these are using the dependency table as was implemented before and no changes are necessary. these providers will not register a provider endpoint
b) services that are not using any of the dependency / tenancy api - these will never be referenced in our dependency table but we want to do a real-time check before we allow the deletion to happen. In this case we never touch the dependency table as these services will never be referenced there.